chainio/dispatcher.go

```diff
@@ -0,0 +1,106 @@
+package chainio
+
+import (
+	"errors"
+	"fmt"
+	"time"
+)
+
+// DefaultProcessBlockTimeout is the timeout value used when waiting for one
+// consumer to finish processing the new block epoch.
+var DefaultProcessBlockTimeout = 60 * time.Second
+
+// ErrProcessBlockTimeout is the error returned when a consumer takes too long
+// to process the block.
+var ErrProcessBlockTimeout = errors.New("process block timeout")
+
+// DispatchSequential takes a list of consumers and notify them about the new
+// epoch sequentially.
+func DispatchSequential(beat Blockbeat, consumers []Consumer) error {
+	for _, c := range consumers {
+		b := beat.copy()
+
+		// Send the copy of the beat to the consumer.
+		err := notifyAndWait(b, c, DefaultProcessBlockTimeout)
+		if err != nil {
+			b.logger().Errorf("Failed to process block: %v", err)
+
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DispatchConcurrent notifies each consumer concurrently about the blockbeat.
+func DispatchConcurrent(beat Blockbeat, consumers []Consumer) error {
+	// errChans is a map of channels that will be used to receive errors
+	// returned from notifying the consumers.
+	errChans := make(map[string]chan error, len(consumers))
+
+	// Notify each queue in goroutines.
+	for _, c := range consumers {
+		// Create a signal chan.
+		errChan := make(chan error, 1)
+		errChans[c.Name()] = errChan
+
+		// Notify each consumer concurrently.
+		go func(c Consumer, beat Blockbeat) {
+			b := beat.copy()
+
+			// Send the copy of the beat to the consumer.
+			errChan <- notifyAndWait(
+				b, c, DefaultProcessBlockTimeout,
+			)
+		}(c, beat)
+	}
+
+	// Wait for all consumers in each queue to finish.
+	for name, errChan := range errChans {
+		err := <-errChan
+		if err != nil {
+			beat.logger().Errorf("Consumer=%v failed to process "+
+				"block: %v", name, err)
+
+			return err
+		}
+	}
+
+	return nil
+}
+
+// notifyAndWait sends the blockbeat to the specified consumer. It requires the
+// consumer to finish processing the block within the specified time, otherwise
+// a timeout error is returned.
+func notifyAndWait(b Blockbeat, c Consumer, timeout time.Duration) error {
+	b.logger().Debugf("Waiting for consumer[%s] to process it", c.Name())
+
+	// Record the time it takes the consumer to process this block.
+	start := time.Now()
+
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- c.ProcessBlock(b)
```

Comment:
as mentioned above, i don't think it's the right move since we already have the quit chan - adding a context here means another state to be managed hence more complexity.

Reply:
ok yeah understand what you mean, so then let's anticipate that we close the quit channel when we timeout in `notifyAndWait` ?

Reply:
I actually don't think it's necessary to use the `errgroup` - it was first used because we thought it would limit the number of goroutines, as suggested [here](https://github.com/lightningnetwork/lnd/pull/8894#discussion_r1677906997). Then Later we found out it's no different than calling `go func()` as everything is managed by the goroutine scheduler, unless we specifically set the limit here via `SetLimit`. However I like that I don't need to initialize a slice of err chans and wait for them in the end, so I used `errgroup` as a shortcut.

On the other hand, when we shut down, a quit signal is received in both `ProcessBlock` and `notifyQueues`, so both the dispatcher and the consumer will exit.